contributing guideline updated to include CI Pipeline for Azure (#673)
gruntwork-io/terratest: docs/_docs/04_community/contributing.md

   test output so we can verify that everything is working.
1. Any notes on backwards incompatibility or downtime.

#### Validate the Pull Request for Azure Platform

> After [creating a pull request](https://help.github.com/articles/creating-a-pull-request/), please follow the below guideline for [Azure Platform](https://azure.com)

Reviewer: ```suggestion
If you're contributing code for the [Azure Platform](https://azure.com), please follow the below guidelines after [creating a pull request](https://help.github.com/articles/creating-a-pull-request/):
```

> ##### To run the CI Pipeline on forked repo

Reviewer: ```suggestion
##### Run the Azure CI Pipeline on your forked repo

We have a separate CI pipeline for Azure code. To run it on a forked repo:

```

Reviewer: Do we expect all Azure contributors to run the tests this way? Or are these instructions specifically for MS employees?

Author: we recommend them to follow this guideline if they have an active Azure subscription.

it'll shorten the develop-test-fail-fix loop.

'cause they can run the pipeline on their repo, fix the issues, create a pr.

it'll be much faster than, create a pr, check the result, fix the issues, update the pr.

> 1. Run the following command on your preffered Terminal and copy the output

Reviewer: ```suggestion
1. Run the following command on your preferred Terminal to create Azure credentials and copy the output:

```

> ```bash
> az ad sp create-for-rbac --name "terratest-az-cli" --role contributor --sdk-auth
> ```

Reviewer: If you indent the code block by 4 spaces, it'll appear properly as a "sub item" of the list.

```suggestion
    ```bash
    az ad sp create-for-rbac --name "terratest-az-cli" --role contributor --sdk-auth
    ```

```

Author: cool, thanks!..

> 2. Go to Secrets page ([https://github.com/<GITHUB_ACCOUNT>/terratest/settings/secrets](https://github.com/<GITHUB_ACCOUNT>/terratest/settings/secrets)) under Settings tab ([https://github.com/<GITHUB_ACCOUNT>/terratest/settings](https://github.com/<GITHUB_ACCOUNT>/terratest/settings)) on GitHub

Reviewer: NIT: to make the numbered list easier to maintain, could you change these all to `1.`? Markdown will automatically replace them with the proper numbers.

```suggestion
1. Go to Secrets page ([https://github.com/<GITHUB_ACCOUNT>/terratest/settings/secrets](https://github.com/<GITHUB_ACCOUNT>/terratest/settings/secrets)) under Settings tab ([https://github.com/<GITHUB_ACCOUNT>/terratest/settings](https://github.com/<GITHUB_ACCOUNT>/terratest/settings)) on GitHub

```

Author: thanks for the suggestion, I just changed list numbers with `1`.

also ran the `MarkdownLinter` and removed couple of empty lines from the file.

> 3. Create a new `Secret` named `AZURE_CREDENTIALS` and paste the output from the 1<sup>st</sup> step as the value

Reviewer: ```suggestion
3. Create a new `Secret` named `AZURE_CREDENTIALS` and paste the Azure credentials you copied from the 1<sup>st</sup> step as the value

```

> 4. [New Personal Access Token](https://github.com/settings/tokens/new) page under [Settings](https://github.com/settings/profile) / [Developer Settings](https://github.com/settings/apps), make sure `write:discussion` is checked, click _Generate token_ button and copy the generated PAT (_Personal Access Token_)

Reviewer: ```suggestion
4. Create a [new Personal Access Token (PAT)](https://github.com/settings/tokens/new) page under [Settings](https://github.com/settings/profile) / [Developer Settings](https://github.com/settings/apps), making sure `write:discussion` is checked. Click the _Generate token_ button and copy the generated PAT.

```

##### Run the Azure CI Pipeline on your forked repo

We have a separate CI pipeline for Azure code. To run it on a forked repo:

1. Run the following command on your preferred Terminal to create Azure credentials and copy the output:

    ```bash
    az ad sp create-for-rbac --name "terratest-az-cli" --role contributor --sdk-auth
    ```

1. Go to Secrets page ([https://github.com/<GITHUB_ACCOUNT>/terratest/settings/secrets](https://github.com/<GITHUB_ACCOUNT>/terratest/settings/secrets)) under Settings tab ([https://github.com/<GITHUB_ACCOUNT>/terratest/settings](https://github.com/<GITHUB_ACCOUNT>/terratest/settings)) on GitHub

1. Create a new `Secret` named `AZURE_CREDENTIALS` and paste the Azure credentials you copied from the 1<sup>st</sup> step as the value

1. Create a [new Personal Access Token (PAT)](https://github.com/settings/tokens/new) page under [Settings](https://github.com/settings/profile) / [Developer Settings](https://github.com/settings/apps), making sure `write:discussion` is checked. Click the _Generate token_ button and copy the generated PAT.

> 1. Create a new `Secret` named `PAT` and paste the output from the 4<sup>th</sup> step as the value

Reviewer 2: ```suggestion
1. In your [fork project settings/secrets](https://github.com/{YOUR_GUTHIB_ACCOUNT}/terratest/settings/secrets), [Create a new Secret](https://docs.github.com/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-a-repository) named `PAT` and paste the output from the 4<sup>th</sup> step as the value

```

1. Go to Actions tab on GitHub ([https://github.com/<GITHUB_ACCOUNT>/terratest/actions](https://github.com/<GITHUB_ACCOUNT>/terratest/actions))

1. Click `ci-workflow` workflow

1. Click `Run workflow` button and fill the fields in the drop down
    * _Repository Info_ : name of the forked repo (_e.g. xyz/terratest_)
    * _Name of the branch_ : branch name on the forked repo (_e.g. feature/adding-some-important-module_)
    * _Name of the official terratest repo_ : home of the target pr (_gruntwork-io/terratest_)

>     * PR number on the official terratest repo : pr number on the official terratest repo (_e.g. 14, 25, etc._)

Reviewer 2: ```suggestion
    * PR number on the official terratest repo : pr number on the official terratest repo (_e.g. 14, 25, etc._).  Setting this value will leave a success/failure comment in the PR once CI completes execution.

```

    * Skip provider registration : set true if you want to skip terraform provider registration for debug purposes (_false_ or _true_)

1. Wait for the `ci-workflow` to be finished

1. PR with the given _PR Number_ will have the result of the `ci-workflow` as a comment

### Merge and release

> The maintainers for this repo will review your code and provide feedback. If everything looks good, they will merge the

Reviewer 2: ```suggestion
The maintainers for this repo will review your code and provide feedback. Once the PR is accepted, they will merge the

```

   test output so we can verify that everything is working.
1. Any notes on backwards incompatibility or downtime.

#### Validate the Pull Request for Azure Platform

> If you're contributing code for the [Azure Platform](https://azure.com), please follow the below guidelines after [creating a pull request](https://help.github.com/articles/creating-a-pull-request/):
> 
> ##### Run the Azure CI Pipeline on your forked repo

Reviewer: ```suggestion
If you're contributing code for the [Azure Platform](https://azure.com), please follow the below guidelines after [creating a pull request](https://help.github.com/articles/creating-a-pull-request/). If you're contributing code for any other platform (e.g., AWS, GCP, etc), you can skip these steps.

```

   test output so we can verify that everything is working.
1. Any notes on backwards incompatibility or downtime.

#### Validate the Pull Request for Azure Platform

If you're contributing code for the [Azure Platform](https://azure.com), please follow the below guidelines after [creating a pull request](https://help.github.com/articles/creating-a-pull-request/):

##### Run the Azure CI Pipeline on your forked repo

We have a separate CI pipeline for Azure code. To run it on a forked repo:

1. Run the following [Azure Cli](https://docs.microsoft.com/cli/azure/) command on your preferred Terminal to create Azure credentials and copy the output:

    ```bash
    az ad sp create-for-rbac --name "terratest-az-cli" --role contributor --sdk-auth
    ```

> 1. Go to Secrets page ([https://github.com/<GITHUB_ACCOUNT>/terratest/settings/secrets](https://github.com/<GITHUB_ACCOUNT>/terratest/settings/secrets)) under Settings tab ([https://github.com/<GITHUB_ACCOUNT>/terratest/settings](https://github.com/<GITHUB_ACCOUNT>/terratest/settings)) on GitHub

Reviewer 2: ```suggestion
1. Go to Secrets settings page under Settings tab, ([https://github.com/<GITHUB_ACCOUNT>/terratest/settings](https://github.com/<GITHUB_ACCOUNT>/terratest/settings)), on GitHub

```

1. Create a new `Secret` named `AZURE_CREDENTIALS` and paste the Azure credentials you copied from the 1<sup>st</sup> step as the value

1. Create a [new Personal Access Token (PAT)](https://github.com/settings/tokens/new) page under [Settings](https://github.com/settings/profile) / [Developer Settings](https://github.com/settings/apps), making sure `write:discussion` scope is checked. Click the _Generate token_ button and copy the generated PAT.

> 

Reviewer 2: remove extra linebreaks (md linter in vscode should show these.

Author: removed, thanks...

> 1. In your [fork project settings/secrets](https://github.com/{YOUR_GUTHIB_ACCOUNT}/terratest/settings/secrets), [Create a new Secret](https://docs.github.com/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-a-repository) named `PAT` and paste the output from the 4<sup>th</sup> step as the value

Reviewer 2: I think this link is redundant.

```suggestion
1. Go back to settings/secrets in your fork and [Create a new Secret](https://docs.github.com/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-a-repository) named `PAT`.  Paste the output from the 4<sup>th</sup> step as the value

```

   test output so we can verify that everything is working.
1. Any notes on backwards incompatibility or downtime.

#### Validate the Pull Request for Azure Platform

If you're contributing code for the [Azure Platform](https://azure.com), please follow the below guidelines after [creating a pull request](https://help.github.com/articles/creating-a-pull-request/):

##### Run the Azure CI Pipeline on your forked repo

We have a separate CI pipeline for Azure code. To run it on a forked repo:

1. Run the following [Azure Cli](https://docs.microsoft.com/cli/azure/) command on your preferred Terminal to create Azure credentials and copy the output:

    ```bash
    az ad sp create-for-rbac --name "terratest-az-cli" --role contributor --sdk-auth
    ```

> 1. Go to Secrets settings page under Settings tab in your forked project, ([https://github.com/<GITHUB_ACCOUNT>/terratest/settings](https://github.com/<GITHUB_ACCOUNT>/terratest/settings)), on GitHub.

Reviewer 2: ```suggestion
1. Go to Secrets settings page under `Settings` tab in your forked project, `https://github.com/<YOUR_GITHUB_ACCOUNT>/terratest/settings`, on GitHub.
```

   test output so we can verify that everything is working.
1. Any notes on backwards incompatibility or downtime.

#### Validate the Pull Request for Azure Platform

If you're contributing code for the [Azure Platform](https://azure.com) and if you have and active _Azure subscription_, it's recommended to follow the below guidelines after [creating a pull request](https://help.github.com/articles/creating-a-pull-request/). If you're contributing code for any other platform (e.g., AWS, GCP, etc), you can skip these steps.

> > Once the PR has `Azure` tag and _Approved_, following pipeline will run automatically on an active Azure subscription, which [Microsoft](https://microsoft.com) provides.

Reviewer: Prob worth calling out that Terratest maintainers will add this tag to automatically run tests for the PR, but if you want to run the tests yourself before then, you can follow the steps below.

Author: I just updated, thanks for the feedback

> > Once the Terratest maintainers add `Azure` tag and _Approve_ the PR, following pipeline will run automatically on an active Azure subscription, which [Microsoft](https://microsoft.com) provides. You can run the pipeline manually on your forked repo by following the below guideline.

Reviewer 2: ```suggestion
> Once the Terratest maintainers add `Azure` tag and _Approve_ the PR, following pipeline will run automatically to perform a full validation of the Azure contribution. You also can run the pipeline manually on your forked repo by following the below guideline.

```

We have a separate CI pipeline for _Azure_ code. To run it on a forked repo:

1. Run the following [Azure Cli](https://docs.microsoft.com/cli/azure/) command on your preferred Terminal to create Azure credentials and copy the output:

    ```bash
    az ad sp create-for-rbac --name "terratest-az-cli" --role contributor --sdk-auth
    ```

1. Go to Secrets settings page under `Settings` tab in your forked project, `https://github.com/<YOUR_GITHUB_ACCOUNT>/terratest/settings`, on GitHub.

1. Create a new `Secret` named `AZURE_CREDENTIALS` and paste the Azure credentials you copied from the 1<sup>st</sup> step as the value

>     > `AZURE_CREDENTIALS` will be stored on _your GitHub account_. GitHub stores your secrets, encrypted an in a very secure place ([GitHub Actions Secrets Reference](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets)). Once the secret created, it's only possible to update or delete it, value of the secrets can't be viewed. GitHub uses a [libsodium sealed box](https://libsodium.gitbook.io/doc/public-key_cryptography/sealed_boxes) to help ensure that secrets are encrypted before they reach GitHub.

Reviewer: ```suggestion
    > `AZURE_CREDENTIALS` will be stored in _your_ GitHub account; neither the Terratest maintainers nor anyone else will have any access to it. Under the hood, GitHub stores your secrets in a secure, encrypted format (see: [GitHub Actions Secrets Reference](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets)). Once the secret is created, it's only possible to update or delete it; the value of the secret can't be viewed. GitHub uses a [libsodium sealed box](https://libsodium.gitbook.io/doc/public-key_cryptography/sealed_boxes) to help ensure that secrets are encrypted before they reach GitHub.
```

1. Create a [new Personal Access Token (PAT)](https://github.com/settings/tokens/new) page under [Settings](https://github.com/settings/profile) / [Developer Settings](https://github.com/settings/apps), making sure `write:discussion` and `public_repo` scopes are checked. Click the _Generate token_ button and copy the generated PAT.

1. Go back to settings/secrets in your fork and [Create a new Secret](https://docs.github.com/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-a-repository) named `PAT`.  Paste the output from the 4<sup>th</sup> step as the value

>     > `PAT` will be stored on _your GitHub account_. GitHub stores your secrets, encrypted an in a very secure place ([GitHub Actions Secrets Reference](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets)). Once the secret created, it's only possible to update or delete it, value of the secrets can't be viewed. GitHub uses a [libsodium sealed box](https://libsodium.gitbook.io/doc/public-key_cryptography/sealed_boxes) to help ensure that secrets are encrypted before they reach GitHub.

Reviewer: ```suggestion
    > `PAT` will be stored in _your_ GitHub account; neither the Terratest maintainers nor anyone else will have any access to it. Under the hood, GitHub stores your secrets in a secure, encrypted format (see: [GitHub Actions Secrets Reference](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets)). Once the secret is created, it's only possible to update or delete it; the value of the secret can't be viewed. GitHub uses a [libsodium sealed box](https://libsodium.gitbook.io/doc/public-key_cryptography/sealed_boxes) to help ensure that secrets are encrypted before they reach GitHub.
    
```

1. Go to Actions tab on GitHub ([https://github.com/<GITHUB_ACCOUNT>/terratest/actions](https://github.com/<GITHUB_ACCOUNT>/terratest/actions))

1. Click `ci-workflow` workflow

1. Click `Run workflow` button and fill the fields in the drop down
    * _Repository Info_ : name of the forked repo (_e.g. xyz/terratest_)
    * _Name of the branch_ : branch name on the forked repo (_e.g. feature/adding-some-important-module_)
    * _Name of the official terratest repo_ : home of the target pr (_gruntwork-io/terratest_)
    * PR number on the official terratest repo : pr number on the official terratest repo (_e.g. 14, 25, etc._).  Setting this value will leave a success/failure comment in the PR once CI completes execution.

    * Skip provider registration : set true if you want to skip terraform provider registration for debug purposes (_false_ or _true_)

1. Wait for the `ci-workflow` to be finished

>     > The pipeline will use the given Azure subscription and create resources on it. When the test finish, resources will be tear-down. Creating resources may cost to the Azure subscription, though. We didn't expect a left-over resource, but please keep in mind that, sometimes due to a bug or a glitch, some resources may be left behind.

Reviewer: ```suggestion
    > The pipeline will use the given Azure subscription and deploy real resources in your Azure account. When the tests finish, they will tear down the resources they created. Of course, if there is a bug or glitch that prevents the clean up code from running, some resources may be left behind, but this is rare. Note that these resources may cost you money! You are responsible for all charges in your Azure subscription.

```

1. Create a new `Secret` named `AZURE_CREDENTIALS` and paste the Azure credentials you copied from the 1<sup>st</sup> step as the value

>     > `AZURE_CREDENTIALS` will be stored on _your GitHub account_. GitHub stores your secrets, encrypted an in a very secure place ([GitHub Actions Secrets Reference](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets)). Once the secret created, it's only possible to update or delete it, value of the secrets can't be viewed. GitHub uses a [libsodium sealed box](https://libsodium.gitbook.io/doc/public-key_cryptography/sealed_boxes) to help ensure that secrets are encrypted before they reach GitHub.

Reviewer 2: ```suggestion
    > `AZURE_CREDENTIALS` will be stored on _your GitHub account_. GitHub stores your secrets, encrypted in a secure location.  See ([GitHub Actions Secrets Reference](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets)) for more detailed information.

```

Author: contributing guideline updated, thanks for the feedback

1. Create a [new Personal Access Token (PAT)](https://github.com/settings/tokens/new) page under [Settings](https://github.com/settings/profile) / [Developer Settings](https://github.com/settings/apps), making sure `write:discussion` and `public_repo` scopes are checked. Click the _Generate token_ button and copy the generated PAT.

1. Go back to settings/secrets in your fork and [Create a new Secret](https://docs.github.com/actions/reference/encrypted-secrets#creating-encrypted-secrets-for-a-repository) named `PAT`.  Paste the output from the 4<sup>th</sup> step as the value

>     > `PAT` will be stored on _your GitHub account_. GitHub stores your secrets, encrypted an in a very secure place ([GitHub Actions Secrets Reference](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets)). Once the secret created, it's only possible to update or delete it, value of the secrets can't be viewed. GitHub uses a [libsodium sealed box](https://libsodium.gitbook.io/doc/public-key_cryptography/sealed_boxes) to help ensure that secrets are encrypted before they reach GitHub.

Reviewer 2: ```suggestion
    > `PAT`will be stored on _your GitHub account_. GitHub stores your secrets, encrypted in a secure location.  See ([GitHub Actions Secrets Reference](https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets)) for more detailed information.

```

Author: contributing guideline updated, thanks for the feedback

1. Wait for the `ci-workflow` to be finished

>     > The pipeline will use the given Azure subscription and create resources on it. When the test finish, resources will be tear-down. Creating resources may cost to the Azure subscription, though. We didn't expect a left-over resource, but please keep in mind that, sometimes due to a bug or a glitch, some resources may be left behind.

Reviewer 2: ```suggestion
    > The pipeline will use the given Azure subscription and create resources as part of running the test in this subscription.  When the test finish the resources are torn down. You will incur some cost for these resources during the execution.  Normal operation should not leave any resource post CI execution, but it can happen.  You should verify that the resources have been deleted post workflow execution.

```

Author: contributing guideline updated, thanks for the feedback
